pkg/benchmarkjunit/parse.go

```diff
@@ -0,0 +1,222 @@
+type Failure struct {
+	Text string `xml:",chardata"`
+}
+
+// TestCase defines the xml element that stores all information associated with one benchmark.
+type TestCase struct {
+	ClassName    string     `xml:"class_name,attr"`
+	Name         string     `xml:"name,attr"`
+	Time         string     `xml:"time,attr"`
+	Failure      *Failure   `xml:"failure,omitempty"`
+	Skipped      bool       `xml:"skipped,omitempty"`
+	PropertyList Properties `xml:"properties"`
+}
+
+// TestSuite defines the outer-most xml element that contains all TestCases.
+type TestSuite struct {
+	Name      string     `xml:"name,attr"`
+	Tests     int        `xml:"tests,attr"`
+	Failures  int        `xml:"failures,attr"`
+	Time      string     `xml:"time,attr"`
+	TestCases []TestCase `xml:"testcase"`
+}
+
+// TestSuites defines the xml element that stores the list of TestSuites
+type TestSuites struct {
+	XMLName string      `xml:"testsuites"`
+	List    []TestSuite `xml:"testsuite"`
+}
+
+func truncate(str string, n int) string {
+	if len(str) <= n {
+		return str
+	}
+	if n > 3 {
+		return str[:n-3] + "..."
+	}
+	return str[:n]
+}
+
+func (s *TestSuite) recordLogText(text string) {
+	if len(s.TestCases) == 0 {
+		logrus.Error("Tried to record Benchmark log text before any Benchmarks were found for the package!")
+		return
+	}
+	tc := &s.TestCases[len(s.TestCases)-1]
+	// TODO: make failure text truncation configurable.
+	text = truncate(text, 1000)
+	// For now only record log text for failures.
+	if tc.Failure == nil {
+		return
+	}
+	tc.Failure.Text = text
+	// Also add failure text to "categorized_fail" property for TestGrid.
+	tc.PropertyList.PropertyList = append(
+		tc.PropertyList.PropertyList,
+		Property{Name: "categorized_fail", Value: text},
+	)
+}
+
+// propertiesFromMatch returns the PropertyList and test duration for a Benchmark metric line match (reBenchMetrics).
+func propertiesFromMatch(match []string) (*Properties, string, error) {
+	var props []Property
+	opCount, err := strconv.ParseFloat(match[2], 64)
+	if err != nil {
+		return nil, "", fmt.Errorf("error parsing opcount %q: %v", match[2], err)
+	}
+	opDuration, err := strconv.ParseFloat(match[3], 64)
+	if err != nil {
+		return nil, "", fmt.Errorf("error parsing ns/op %q: %v", match[3], err)
+	}
+	benchmarkDuration := fmt.Sprintf("%f", opCount*opDuration/1000000000) // convert from ns to s.
+
+	props = append(props, Property{Name: "op count", Value: match[2]})
+	props = append(props, Property{Name: "avg op duration (ns/op)", Value: match[3]})
+	if len(match[4]) > 0 {
+		props = append(props, Property{Name: "MB/s", Value: match[4]})
+	}
+	if len(match[5]) > 0 {
+		props = append(props, Property{Name: "alloced B/op", Value: match[5]})
+	}
+	if len(match[6]) > 0 {
+		props = append(props, Property{Name: "allocs/op", Value: match[6]})
+	}
+
+	return &Properties{PropertyList: props}, benchmarkDuration, nil
+}
+
+func parse(raw []byte) (*TestSuites, error) {
```

Comment:
Since you have the testdata dir, we could write an integration test if you wanted